README


2PING - A BI-DIRECTIONAL PING UTILITY


https://www.finnie.org/software/2ping/


About

2ping is a bi-directional ping utility. It uses 3-way pings (akin to TCP
SYN, SYN/ACK, ACK) and after-the-fact state comparison between a 2ping
listener and a 2ping client to determine which direction packet loss
occurs.


Installation

2ping requires Python 3 version 3.4 or higher.

To install:

    sudo python3 setup.py install

Python 3 stdlib is the only requirement for base functionality, but
2ping can utilize the following modules if available:

-   dnspython for --srv
-   netifaces for --listen --all-interfaces
-   pycrypto for --encrypt


Usage

Please see the 2ping manpage for invocation options, but in short, start
a listener on the far end:

    2ping --listen

And run 2ping on the near end, connecting to the far end listener:

    2ping $LISTENER

Where "$LISTENER" is the name or IP address of the listener.


License

2ping - A bi-directional ping utility

Copyright (C) 2010-2018 Ryan Finnie

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or (at your
option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.


ChangeLog
2ping 4.5 (2020-06-18)
 * Added PyCryptodome support (recommended over PyCrypto, though the latter
   is still detected/supported).
 * Replaced best_poller module with Python native selectors module.
 * Changed --flood output: dots/backspaces are no longer printed, and loss
   results / errors display full details.
 * --audible tones will only occur if stdout is a TTY.
 * Improved hostname/IP display edge cases.
 * Added an AF_UNIX --loopback test mode.
 * Listener sockets are added and removed as needed, instead of being
   re-created on each rescan.
 * Listener sockets are automatically rescanned periodically.
 * Multiple systemd sockets are now allowed.
 * A run can be both a listener and a client at the same time (mainly
   useful for smoke testing).
 * Other socket handling refactoring.
 * Other code refactoring.
 * Listener statistics are displayer per-bind.
 * Many, many testing/CI improvements.

2ping 4.4.1 (2020-06-08)
 * Fixed 2ping.spec referencing old README and making
   `rpmbuild -ta 2ping.tar.gz` fail.
 * Added systemd 2ping.service.
 * Added snapcraft.yaml.

2ping 4.4 (2020-06-07)
 * Minimum Python version changed from 3.4 to 3.5
 * Monotonic clock is always used
 * If the Python "netifaces" module is installed (preferred), --listen
   will now listen on all addresses by default, as opposed to requiring
   --all-interfaces previously
 * Add --subtract-peer-host-latency
 * Add support for systemd-supplied sockets
 * Remove deprecated/removed linux_distribution, use distro if available
 * Code/documentation cleanups, 2ping protocol 4.1

2ping 4.3 (2018-12-03)
 * Add --srv-service
 * Change --adaptive behavior to better match ping -A
 * Fix typos in manpage

2ping 4.2 (2018-08-11)
 * Added SIGHUP handling of listener processes
 * Added an example bash_completion script
 * Better cleanup handling of peer information

2ping 4.1.2 (2018-08-09)
 * Fix UTF-8 tests when run with invalid locale (Debian Bug#897498)
 * Fix cleanup on non-encrypted sessions (GitHub rfinnie/2ping#5)

2ping 4.1 (2017-08-06)
 * Fixed --fuzz CRC function.
 * Added --encrypt option for shared-secret encrypted packets.
 * Added --listen --all-interfaces option for automatically binding to
   all interface IPs (requires Python netifaces module).
 * Simplified monotonic_clock functionality, relying on Python 3 for
   most functionality, reducing the possibility of platform bugs.
 * Minor fixes and unit test suite improvements.

2ping 4.0.1 (2017-07-22)
 * Fixed unit tests causing test failure in certain conditions.

2ping 4.0 (2017-07-22)
 * Rewrite from Python 2 to Python 3 (3.4 or higher).
 * Fixed hmac-sha256 handling, added hmac-sha512.
 * --nagios will now work when combined with --deadline, in addition to
   --count.
 * Added Wireshark Lua dissector and sample capture.
 * Added battery level (ExtID 0x88a1f7c7).  Note that while 2ping
   recognizes the new option in incoming packets, it currently does not
   have the capability to send battery levels.
 * Minor fixes.

2ping 3.2.1 (2016-03-26)
  * Do not error out when non-ASCII notice text is received (only causes
    a remote denial of service crash when --debug is specified on the
    remote peer).

2ping 3.2.0 (2016-02-10)
  * Added --nagios, for Nagios-compatible output and status codes.
  * Added unit tests.
  * Added --send-time, which sends an extended segment containing the
    current wall time.
  * Added --send-monotonic-clock, which sends an extended segment
    containing a monotonically-incrementing counter, on supported platforms.
  * Added --send-random, which sends an extended segment containing
    random bytes.
  * Added -fuzz, which randomly fuzzes incoming packets (developer
    feature).
  * Fixed over-cautious handling of length limits when assembling
    extended segments.

2ping 3.1.0 (2015-11-16)
  * Best available poller for each platform (e.g. epoll on Linux, kqueue
    on BSD / OS X) is automatically used.
  * Old age timeout is set to a lower value on Win32 (1 second instead
    of 60), as KeyboardInterrupt does not interrupt select() on Win32.
  * Packet loss is now better visible in flood mode.
  * Adaptive mode now ramps up to EWMA faster.
  * Adaptive mode RTT predictions are now calculated per destination.
  * In client mode, statistics are now separated for each destination.
  * Added optional DNS SRV client support (requires dnspython).  When
    given --srv, all SRV records for the 2ping UDP service of a host are
    pinged in parallel.
  * Investigation results are now sorted by sequence number.
  * Hostnames are displayed in statistics, if known.
  * 2ping will exit earlier if safe to do so (e.g. "-c 1" will not wait
    a full second if the ping is received immediately).
  * --port can now be given service names (as determined by the system
    resolver) instead of numeric ports.
  * System platform (Linux, Mach, etc) is sent in packets along with
    2ping version.
  * Statistics use a more human-readable format (m, s, ms, etc).

2ping 3.0.1 (2015-10-29)
  * Fix peer_address on error when MSG_ERRQUEUE is not set
  * Documentation update

2ping 3.0.0 (2015-10-25)
  * Total rewrite from Perl to Python.
  * Multiple hostnames/addresses may be specified in client mode, and
    will be pinged in parallel.
  * Improved IPv6 support:
    * In most cases, specifying -4 or -6 is unnecessary. You should be
      able to specify IPv4 and/or IPv6 addresses and it will "just work".
    * IPv6 addresses may be specified without needing to add -6.
    * If a hostname is given in client mode and the hostname provides
      both AAAA and A records, the AAAA record will be chosen. This can
      be forced to one or another with -4 or -6.
    * If a hostname is given in listener mode with -I, it will be
      resolved to addresses to bind as. If the hostname provides both
      AAAA and A records, they will both be bound. Again, -4 or -6 can
      be used to restrict the bind.

twoping/args.py
# 2ping - A bi-directional ping utility
# Copyright (C) 2010-2020 Ryan Finnie
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301, USA.

import argparse
import os
import socket
import sys

from . import __version__
from .utils import _, AES


def parse_args(argv=None):
    if argv is None:
        argv = sys.argv

    if argv[0].endswith("2ping6"):
        ipv6_default = True
    else:
        ipv6_default = False

    parser = argparse.ArgumentParser(
        description="2ping ({})".format(__version__),
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
        prog=os.path.basename(argv[0]),
    )
    parser.add_argument(
        "--version",
        "-V",
        action="version",
        version=__version__,
        help=_("report the program version"),
    )

    # Positionals
    parser.add_argument(
        "host", type=str, default=None, nargs="*", help=_("host to ping")
    )

    # ping-compatible options
    parser.add_argument(
        "--audible", "-a", dest="audible", action="store_true", help=_("audible ping")
    )
    parser.add_argument(
        "--adaptive",
        "-A",
        dest="adaptive",
        action="store_true",
        help=_("adaptive RTT ping"),
    )
    parser.add_argument(
        "--count", "-c", dest="count", type=int, help=_("number of pings to send")
    )
    parser.add_argument(
        "--flood", "-f", dest="flood", action="store_true", help=_("flood mode")
    )
    parser.add_argument(
        "--interval",
        "-i",
        dest="interval",
        type=float,
        default=1.0,
        help=_("seconds between pings"),
        metavar="SECONDS",
    )
    parser.add_argument(
        "--interface-address",
        "-I",
        dest="interface_address",
        type=str,
        action="append",
        default=[],
        help=_("interface bind address"),
        metavar="ADDRESS",
    )
    parser.add_argument(
        "--preload",
        "-l",
        dest="preload",
        type=int,
        default=1,
        help=_("number of pings to send at start"),
        metavar="COUNT",
    )
    parser.add_argument(
        "--pattern",
        "-p",
        dest="pattern",
        type=str,
        help=_("hex pattern for padding"),
        metavar="HEX_BYTES",
    )
    parser.add_argument(
        "--quiet", "-q", dest="quiet", action="store_true", help=_("quiet mode")
    )
    parser.add_argument(
        "--packetsize-compat",
        "-s",
        dest="packetsize_compat",
        type=int,
        help=_("packet size (ping compatible)"),
        metavar="BYTES",
    )
    parser.add_argument(
        "--verbose", "-v", dest="verbose", action="store_true", help=_("verbose mode")
    )
    parser.add_argument(
        "--deadline",
        "-w",
        dest="deadline",
        type=float,
        help=_("maximum run time"),
        metavar="SECONDS",
    )

    # 2ping options
    parser.add_argument("--all-interfaces", action="store_true", help=_("deprecated"))
    parser.add_argument(
        "--auth", type=str, help=_("HMAC authentication key"), metavar="KEY"
    )
    parser.add_argument(
        "--auth-digest",
        type=str,
        default="hmac-md5",
        choices=["hmac-md5", "hmac-sha1", "hmac-sha256", "hmac-crc32", "hmac-sha512"],
        help=_("HMAC authentication digest"),
        metavar="DIGEST",
    )
    parser.add_argument("--debug", action="store_true", help=_("debug mode"))
    parser.add_argument("--encrypt", type=str, help=_("Encryption key"), metavar="KEY")
    parser.add_argument(
        "--encrypt-method",
        type=str,
        default="hkdf-aes256-cbc",
        choices=["hkdf-aes256-cbc"],
        help=_("Encryption method"),
        metavar="METHOD",
    )
    parser.add_argument(
        "--fuzz", type=float, help=_("incoming fuzz percentage"), metavar="PERCENT"
    )
    parser.add_argument(
        "--inquire-wait",
        type=float,
        default=10.0,
        help=_("maximum time before loss inquiries"),
        metavar="SECONDS",
    )
    parser.add_argument("--ipv4", "-4", action="store_true", help=_("force IPv4"))
    parser.add_argument(
        "--ipv6", "-6", action="store_true", default=ipv6_default, help=_("force IPv6")
    )
    parser.add_argument("--listen", action="store_true", help=_("listen mode"))
    parser.add_argument(
        "--loopback", action="store_true", help=_("UNIX loopback test mode")
    )
    parser.add_argument(
        "--loopback-pairs",
        type=int,
        default=1,
        help=_("number of loopback pairs to create"),
        metavar="PAIRS",
    )
    parser.add_argument(
        "--max-packet-size",
        type=int,
        default=512,
        help=_("maximum packet size"),
        metavar="BYTES",
    )
    parser.add_argument(
        "--min-packet-size",
        type=int,
        default=128,
        help=_("minimum packet size"),
        metavar="BYTES",
    )
    parser.add_argument(
        "--nagios",
        type=str,
        help=_("nagios-compatible output"),
        metavar="WRTA,WLOSS%,CRTA,CLOSS%",
    )
    parser.add_argument(
        "--no-3way", action="store_true", help=_("do not send 3-way pings")
    )
    parser.add_argument(
        "--no-match-packet-size",
        action="store_true",
        help=_("do not match packet size of peer"),
    )
    parser.add_argument(
        "--no-send-version",
        action="store_true",
        help=_("do not send program version to peers"),
    )
    parser.add_argument(
        "--notice", type=str, help=_("arbitrary notice text"), metavar="TEXT"
    )
    parser.add_argument(
        "--packet-loss",
        type=str,
        help=_("percentage simulated packet loss"),
        metavar="OUT:IN",
    )
    parser.add_argument(
        "--port", type=str, default="15998", help=_("port to connect / bind to")
    )
    parser.add_argument(
        "--send-monotonic-clock",
        action="store_true",
        help=_("send monotonic clock to peers"),
    )
    parser.add_argument(
        "--send-random", type=int, help=_("send random data to peers"), metavar="BYTES"
    )
    parser.add_argument(
        "--send-time", action="store_true", help=_("send wall clock time to peers")
    )
    parser.add_argument(
        "--stats", type=float, help=_("print recurring statistics"), metavar="SECONDS"
    )
    parser.add_argument(
        "--srv", action="store_true", help=_("lookup SRV records in client mode")
    )
    parser.add_argument(
        "--srv-service",
        type=str,
        default="2ping",
        help=_("service name for SRV lookups"),
    )
    parser.add_argument(
        "--subtract-peer-host-latency",
        action="store_true",
        help=_("subtract peer host latency from RTT calculations, if sent"),
    )

    # ping-compatible ignored options
    for opt in "b|B|d|L|n|R|r|U".split("|"):
        parser.add_argument(
            "-{}".format(opt),
            action="store_true",
            dest="ignored_{}".format(opt),
            help=argparse.SUPPRESS,
        )
    for opt in "F|Q|S|t|T|M|W".split("|"):
        parser.add_argument(
            "-{}".format(opt),
            type=str,
            default=None,
            dest="ignored_{}".format(opt),
            help=argparse.SUPPRESS,
        )

    args = parser.parse_args(args=argv[1:])

    if (not args.listen) and (not args.host) and (not args.loopback):
        parser.print_help()
        parser.exit()
    if args.loopback and not hasattr(socket, "AF_UNIX"):
        parser.error("--loopback not supported on non-UNIX platforms")
    if args.nagios:
        args.quiet = True
        if (not args.count) and (not args.deadline):
            args.count = 5
        nagios_opts = args.nagios.split(",")
        if len(nagios_opts) != 4:
            parser.error(_("Invalid limits"))
        (
            args.nagios_warn_rta,
            args.nagios_warn_loss,
            args.nagios_crit_rta,
            args.nagios_crit_loss,
        ) = nagios_opts
        if args.nagios_warn_loss[-1:] != "%":
            parser.error(_("Invalid limits"))
        if args.nagios_crit_loss[-1:] != "%":
            parser.error(_("Invalid limits"))
        try:
            args.nagios_warn_loss = float(args.nagios_warn_loss[:-1])
            args.nagios_crit_loss = float(args.nagios_crit_loss[:-1])
            args.nagios_warn_rta = float(args.nagios_warn_rta)
            args.nagios_crit_rta = float(args.nagios_crit_rta)
        except ValueError as e:
            parser.error(e.message)
    if args.packetsize_compat:
        args.min_packet_size = args.packetsize_compat + 8
    if args.max_packet_size < args.min_packet_size:
        parser.error(_("Maximum packet size must be at least minimum packet size"))
    if args.max_packet_size < 64:
        parser.error(_("Maximum packet size must be at least 64"))
    args.packet_loss_in = 0
    args.packet_loss_out = 0
    if args.packet_loss:
        if ":" in args.packet_loss:
            (v_out, v_in) = args.packet_loss.split(":", 1)
            try:
                args.packet_loss_out = float(v_out)
                args.packet_loss_in = float(v_in)
            except ValueError as e:
                parser.error(e.message)
        else:
            try:
                args.packet_loss_in = args.packet_loss_out = float(args.packet_loss)
            except ValueError as e:
                parser.error(_("Packet loss: {error}").format(error=e.message))
    if args.pattern:
        if len(args.pattern) & 1:
            parser.error(_("Pattern must be full bytes"))
        if len(args.pattern) > 32:
            parser.error(_("Pattern must be 16 bytes or less"))
        args.pattern_bytes = b""
        for i in range(int(len(args.pattern) / 2)):
            a = args.pattern[(i * 2) : (i * 2 + 2)]
            try:
                b = int(a, 16)
            except ValueError as e:
                parser.error(_("Pattern: {error}").format(error=e.message))
            args.pattern_bytes += bytes([b])
    else:
        args.pattern_bytes = b"\x00"

    hmac_id_map = {
        "hmac-md5": 1,
        "hmac-sha1": 2,
        "hmac-sha256": 3,
        "hmac-crc32": 4,
        "hmac-sha512": 5,
    }
    args.auth_digest_index = hmac_id_map[args.auth_digest]

    if args.encrypt and isinstance(AES, ImportError):
        parser.error(_("Crypto module required for encryption"))

    encrypt_id_map = {"hkdf-aes256-cbc": 1}
    args.encrypt_method_index = encrypt_id_map[args.encrypt_method]

    if args.debug:
        args.verbose = True

    return args
